master/internal/kubernetes/queue.go

```diff
@@ -0,0 +1,215 @@
+package kubernetes
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+
+	k8sV1 "k8s.io/api/core/v1"
+	typedV1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const (
+	numKubernetesWorkers = 5
+	deletionGracePeriod  = 15
+)
+
+// message types that are sent to the requestQueue.
+type (
+	createKubernetesResources struct {
+		handler       *actor.Ref
+		podSpec       *k8sV1.Pod
+		configMapSpec *k8sV1.ConfigMap
+	}
+
+	deleteKubernetesResources struct {
+		handler       *actor.Ref
+		podName       string
+		configMapName string
+	}
+)
+
+// message types that are sent by requestQueue and requestProcessingWorkers as responses
+// to creation or deletion requests.
+type (
+	resourceCreationFailed struct {
+		err error
+	}
+
+	resourceDeletionFailed struct {
+		err error
+	}
+
+	resourceCreationCancelled struct{}
+)
+
+// message types sent from requestProcessingWorkers to requestQueue.
+type (
+	workerAvailable struct{}
+)
+
+// queuedResourceRequest is used to represent requests that are being buffered by requestQueue.
+type queuedResourceRequest struct {
+	createResources *createKubernetesResources
+	deleteResources *deleteKubernetesResources
+}
+
+// The requestQueue is responsible for fulfilling all requests for creating and deleting
+// kubernetes resources that require interaction with the kubernetes API. It accomplishes
+// this by forwarding requests to requestProcessingWorker actors which prcess the request.
+// There are two reasons a queue system is required as opposed to allowing the pod actors
+// to create and delete Kubernetes resources asynchronously:
```

Comment:
question: By "asynchronously", do you mean between different pod actors?

Reply:
yep, i tweaked this comment to make that more clear.